File: scripts/harvest_ecobase_models.py
import requests
import pandas as pd
import json
from pathlib import Path
import logging

# Set up logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class EcoBaseHarvester:
    """Class to harvest EwE model parameters from EcoBase database"""
    
    def __init__(self):
        self.base_url = "https://sirs.agrocampus-ouest.fr/EcoBase/api/v1"
        self.output_dir = Path("data/ecobase_models")
        self.output_dir.mkdir(parents=True, exist_ok=True)
        
    def get_model_list(self):
        """Retrieve list of available models from EcoBase"""
        try:
            response = requests.get(f"{self.base_url}/models")
            if response.status_code == 200:
                models = response.json()
                df = pd.DataFrame(models)
                df.to_csv(self.output_dir / "model_list.csv", index=False)
                logger.info(f"Retrieved {len(models)} models")
                return df
            else:
                logger.error(f"Failed to get model list: {response.status_code}")
                return None
        except Exception as e:
            logger.error(f"Error getting model list: {str(e)}")
            return None

    def get_model_parameters(self, model_id):
        """Retrieve parameters for a specific model"""
        try:
            response = requests.get(f"{self.base_url}/models/{model_id}/parameters")
            if response.status_code == 200:
                params = response.json()
                
                # Save raw parameters
                with open(self.output_dir / f"model_{model_id}_params.json", 'w') as f:
                    json.dump(params, f, indent=2)
                
                # Extract key parameters into structured format
                structured_params = {
                    'biomass': [],
                    'pb_ratio': [],  # Production/Biomass ratio
                    'qb_ratio': [],  # Consumption/Biomass ratio
                    'ee': [],        # Ecotrophic efficiency
                    'diet_matrix': []
                }
                
                # Process parameters based on their type
                for param in params:
                    param_type = param.get('type')
                    if param_type in structured_params:
                        structured_params[param_type].append({
                            'group': param.get('group'),
                            'value': param.get('value'),
                            'unit': param.get('unit')
                        })
                
                # Convert to DataFrames and save
                for param_type, data in structured_params.items():
                    if data:
                        df = pd.DataFrame(data)
                        df.to_csv(self.output_dir / f"model_{model_id}_{param_type}.csv", 
                                index=False)
                
                logger.info(f"Successfully retrieved parameters for model {model_id}")
                return structured_params
            else:
                logger.error(f"Failed to get parameters for model {model_id}: {response.status_code}")
                return None
        except Exception as e:
            logger.error(f"Error getting parameters for model {model_id}: {str(e)}")
            return None

    def search_models(self, ecosystem_type=None, region=None, year=None):
        """Search for models based on criteria"""
        try:
            params = {
                'ecosystem_type': ecosystem_type,
                'region': region,
                'year': year
            }
            # Remove None values
            params = {k: v for k, v in params.items() if v is not None}
            
            response = requests.get(f"{self.base_url}/models/search", params=params)
            if response.status_code == 200:
                models = response.json()
                df = pd.DataFrame(models)
                search_desc = "_".join(f"{k}_{v}" for k, v in params.items())
                if search_desc:
                    output_file = f"model_search_{search_desc}.csv"
                else:
                    output_file = "model_search_all.csv"
                df.to_csv(self.output_dir / output_file, index=False)
                logger.info(f"Found {len(models)} models matching criteria")
                return df
            else:
                logger.error(f"Search failed: {response.status_code}")
                return None
        except Exception as e:
            logger.error(f"Error searching models: {str(e)}")
            return None

def main():
    harvester = EcoBaseHarvester()
    
    # Get list of all models
    models_df = harvester.get_model_list()
    if models_df is not None:
        logger.info("Retrieved model list successfully")
        
        # Example: Search for marine models in a specific region
        marine_models = harvester.search_models(
            ecosystem_type="Marine",
            region="Pacific Ocean"
        )
        
        if marine_models is not None:
            # Get parameters for first 5 marine models
            for model_id in marine_models['model_id'].head():
                params = harvester.get_model_parameters(model_id)
                if params:
                    logger.info(f"Retrieved parameters for model {model_id}")

if __name__ == "__main__":
    main()

File: scripts/harvest_ecobase_models.R
library(httr)
library(jsonlite)
library(dplyr)
library(tidyr)
library(readr)

#' EcoBase Harvester Class
#' @description R6 class to harvest EwE model parameters from EcoBase database
EcoBaseHarvester <- R6::R6Class(
  "EcoBaseHarvester",
  public = list(
    base_url = "https://sirs.agrocampus-ouest.fr/EcoBase/api/v1",
    output_dir = "data/ecobase_models",
    
    initialize = function() {
      dir.create(self$output_dir, recursive = TRUE, showWarnings = FALSE)
    },
    
    #' Get list of available models
    get_model_list = function() {
      tryCatch({
        response <- GET(paste0(self$base_url, "/models"))
        if (status_code(response) == 200) {
          models <- fromJSON(rawToChar(response$content))
          models_df <- as.data.frame(models)
          write_csv(models_df, file.path(self$output_dir, "model_list.csv"))
          message(sprintf("Retrieved %d models", nrow(models_df)))
          return(models_df)
        } else {
          warning(sprintf("Failed to get model list: %d", status_code(response)))
          return(NULL)
        }
      }, error = function(e) {
        warning(sprintf("Error getting model list: %s", e$message))
        return(NULL)
      })
    },
    
    #' Get parameters for a specific model
    get_model_parameters = function(model_id) {
      tryCatch({
        response <- GET(sprintf("%s/models/%s/parameters", self$base_url, model_id))
        if (status_code(response) == 200) {
          params <- fromJSON(rawToChar(response$content))
          
          # Save raw parameters
          write_json(params, file.path(self$output_dir, 
                                     sprintf("model_%s_params.json", model_id)),
                    pretty = TRUE)
          
          # Create structured format for parameters
          structured_params <- list(
            biomass = list(),
            pb_ratio = list(),
            qb_ratio = list(),
            ee = list(),
            diet_matrix = list()
          )
          
          # Process parameters
          for (param in params) {
            param_type <- param$type
            if (!is.null(param_type) && param_type %in% names(structured_params)) {
              structured_params[[param_type]][[length(structured_params[[param_type]]) + 1]] <- 
                list(
                  group = param$group,
                  value = param$value,
                  unit = param$unit
                )
            }
          }
          
          # Convert to data frames and save
          for (param_type in names(structured_params)) {
            if (length(structured_params[[param_type]]) > 0) {
              df <- do.call(rbind, lapply(structured_params[[param_type]], as.data.frame))
              write_csv(df, file.path(self$output_dir,
                                    sprintf("model_%s_%s.csv", model_id, param_type)))
            }
          }
          
          message(sprintf("Successfully retrieved parameters for model %s", model_id))
          return(structured_params)
        } else {
          warning(sprintf("Failed to get parameters for model %s: %d", 
                        model_id, status_code(response)))
          return(NULL)
        }
      }, error = function(e) {
        warning(sprintf("Error getting parameters for model %s: %s", 
                      model_id, e$message))
        return(NULL)
      })
    },
    
    #' Search for models based on criteria
    search_models = function(ecosystem_type = NULL, region = NULL, year = NULL) {
      tryCatch({
        params <- list(
          ecosystem_type = ecosystem_type,
          region = region,
          year = year
        )
        params <- params[!sapply(params, is.null)]
        
        response <- GET(paste0(self$base_url, "/models/search"),
                       query = params)
        
        if (status_code(response) == 200) {
          models <- fromJSON(rawToChar(response$content))
          models_df <- as.data.frame(models)
          
          # Create filename based on search criteria
          search_desc <- paste(names(params), params, sep = "_", collapse = "_")
          output_file <- if (search_desc != "") {
            sprintf("model_search_%s.csv", search_desc)
          } else {
            "model_search_all.csv"
          }
          
          write_csv(models_df, file.path(self$output_dir, output_file))
          message(sprintf("Found %d models matching criteria", nrow(models_df)))
          return(models_df)
        } else {
          warning(sprintf("Search failed: %d", status_code(response)))
          return(NULL)
        }
      }, error = function(e) {
        warning(sprintf("Error searching models: %s", e$message))
        return(NULL)
      })
    }
  )
)

# Example usage
main <- function() {
  harvester <- EcoBaseHarvester$new()
  
  # Get list of all models
  models_df <- harvester$get_model_list()
  if (!is.null(models_df)) {
    message("Retrieved model list successfully")
    
    # Example: Search for marine models in a specific region
    marine_models <- harvester$search_models(
      ecosystem_type = "Marine",
      region = "Pacific Ocean"
    )
    
    if (!is.null(marine_models)) {
      # Get parameters for first 5 marine models
      for (model_id in head(marine_models$model_id, 5)) {
        params <- harvester$get_model_parameters(model_id)
        if (!is.null(params)) {
          message(sprintf("Retrieved parameters for model %s", model_id))
        }
      }
    }
  }
}

if (interactive()) {
  main()
}
